File: languages/python/src/slop_guard/rules/smells/no_large_class.py
"""Reject classes that carry more than one job's worth of members."""

from __future__ import annotations

import ast

from ...rule import Rule

DEFAULT_MAX_METHODS = 10
DEFAULT_MAX_ATTRIBUTES = 10


class NoLargeClass(Rule):
    name = "no-large-class"
    group = "smells"
    reference = "https://refactoring.guru/smells/large-class"
    description = (
        "Disallow classes with more than `max_methods` methods or `max_attributes` attributes, "
        "10 of each by default."
    )

    def visit_ClassDef(self, node: ast.ClassDef) -> None:
        method_limit = self.threshold("max_methods", DEFAULT_MAX_METHODS)
        attribute_limit = self.threshold("max_attributes", DEFAULT_MAX_ATTRIBUTES)
        methods = [
            statement
            for statement in node.body
            if isinstance(statement, (ast.FunctionDef, ast.AsyncFunctionDef))
        ]
        attributes = self._attributes(node)
        if len(methods) > method_limit:
            self.report(
                node,
                f"`{node.name}` declares {len(methods)} methods, past the {method_limit} this "
                f"project allows. A class this wide has more than one reason to change; split the "
                f"group of methods that share data into their own class.",
            )
        elif len(attributes) > attribute_limit:
            self.report(
                node,
                f"`{node.name}` holds {len(attributes)} attributes, past the {attribute_limit} "
                f"this project allows. Group the ones that change together into a smaller type.",
            )
        self.generic_visit(node)

    @staticmethod
    def _attributes(node: ast.ClassDef) -> set[str]:
        found: set[str] = set()
        for statement in node.body:
            if isinstance(statement, ast.AnnAssign) and isinstance(statement.target, ast.Name):
                found.add(statement.target.id)
            elif isinstance(statement, ast.Assign):
                found.update(
                    target.id for target in statement.targets if isinstance(target, ast.Name)
                )
        for child in ast.walk(node):
            if (
                isinstance(child, ast.Attribute)
                and isinstance(child.ctx, ast.Store)
                and isinstance(child.value, ast.Name)
                and child.value.id == "self"
            ):
                found.add(child.attr)
        return found

File: languages/typescript/src/smells/rules/no-large-class.ts
import { defineRule } from "@oxlint/plugins";
import type { ESTree } from "@oxlint/plugins";

import { numberOption, thresholdSchema } from "../shared/options.ts";
import { walk } from "../shared/traversal.ts";

const DEFAULT_MAX_METHODS = 10;
const DEFAULT_MAX_FIELDS = 10;

function className(node: ESTree.ClassBody): string {
	const parent = node.parent;
	if (
		(parent?.type === "ClassDeclaration" || parent?.type === "ClassExpression") &&
		parent.id !== null &&
		parent.id !== undefined
	) {
		return parent.id.name;
	}
	return "This class";
}

/** Ban classes that carry more than one job's worth of members. */
export const noLargeClassRule = defineRule({
	meta: {
		type: "problem",
		docs: {
			description:
				"Disallow classes with more than `maxMethods` methods or `maxFields` fields; a class this wide has more than one reason to change.",
		},
		messages: {
			manyMethods:
				"`{{name}}` declares {{count}} methods, past the {{limit}} this project allows. A class this wide has more than one reason to change; split the group of methods that share data into their own class.",
			manyFields:
				"`{{name}}` holds {{count}} fields, past the {{limit}} this project allows. Group the ones that change together into a smaller type.",
		},
		...thresholdSchema({ maxMethods: DEFAULT_MAX_METHODS, maxFields: DEFAULT_MAX_FIELDS }),
	},
	createOnce(context) {
		return {
			ClassBody(node) {
				const methodLimit = numberOption(context.options, "maxMethods", DEFAULT_MAX_METHODS);
				const fieldLimit = numberOption(context.options, "maxFields", DEFAULT_MAX_FIELDS);
				const methods = node.body.filter((member) => member.type === "MethodDefinition");
				const fields = new Set<string>();
				for (const member of node.body) {
					if (member.type === "PropertyDefinition" && member.key.type === "Identifier") {
						fields.add(member.key.name);
					}
				}
				walk(node, context.sourceCode.visitorKeys, (child) => {
					if (
						child.type === "MemberExpression" &&
						child.object.type === "ThisExpression" &&
						child.property.type === "Identifier" &&
						child.parent?.type === "AssignmentExpression"
					) {
						fields.add(child.property.name);
					}
				});
				if (methods.length > methodLimit) {
					context.report({
						node,
						messageId: "manyMethods",
						data: {
							name: className(node),
							count: String(methods.length),
							limit: String(methodLimit),
						},
					});
					return;
				}
				if (fields.size > fieldLimit) {
					context.report({
						node,
						messageId: "manyFields",
						data: { name: className(node), count: String(fields.size), limit: String(fieldLimit) },
					});
				}
			},
		};
	},
});
